plugin/app/feishu/feishu.go

```diff
@@ -0,0 +1,426 @@
+	} `json:"data"`
+}
+
+type user struct {
+	UserID string `json:"user_id"`
+	Email  string `json:"email"`
+}
+
+// GetExternalApprovalResponse is the response of GetExternalApproval.
+type GetExternalApprovalResponse struct {
+	Code int    `json:"code"`
+	Msg  string `json:"msg"`
+	Data struct {
+		Status string `json:"status"`
+	} `json:"data"`
+}
+
+// CancelExternalApprovalResponse is the response of CancelExternalApproval.
+type CancelExternalApprovalResponse struct {
+	Code int      `json:"code"`
+	Msg  string   `json:"msg"`
+	Data struct{} `json:"data"`
+}
+
+type createApprovalInstanceReq struct {
+	ApprovalCode           string `json:"approval_code"`
+	Form                   string `json:"form"`
+	NodeApproverOpenIDList []struct {
+		Key   string   `json:"key"`
+		Value []string `json:"value"`
+	} `json:"node_approver_open_id_list"`
+	OpenID string `json:"open_id"`
+}
+
+// Content is the content of the approval.
+type Content struct {
+	Issue       string
+	Stage       string
+	Description string
+}
+
+func formatForm(content Content) (string, error) {
+	type form []struct {
+		ID    string `json:"id"`
+		Type  string `json:"type"`
+		Value string `json:"value"`
+	}
+	forms := form{
+		{
+			ID:    "1",
+			Type:  "input",
+			Value: content.Issue,
+		},
+		{
+			ID:    "2",
+			Type:  "input",
+			Value: content.Stage,
+		},
+		{
+			ID:    "3",
+			Type:  "textarea",
+			Value: content.Description,
+		},
+	}
+	b, err := json.Marshal(forms)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+const (
+	getTenantAccessTokenReq = `{
+		"app_id": "%s",
+		"app_secret": "%s"
+}`
+	createApprovalDefinitionReq = `{
+  "approval_code": "%s",
+	"approval_name": "@i18n@approval_name",
+	"form": {
+		"form_content": "[{\"id\":\"1\", \"type\": \"input\", \"name\":\"@i18n@widget1\"},{\"id\":\"2\", \"type\": \"input\", \"name\":\"@i18n@widget2\"},{\"id\":\"3\", \"type\": \"textarea\", \"name\":\"@i18n@widget3\"}]"
+	},
+	"i18n_resources": [
+		{
+			"is_default": "true",
+			"locale": "zh-CN",
+			"texts": [
+				{
+					"key": "@i18n@approval_name",
+					"value": "Bytebase 工单"
+				},
+				{
+					"key": "@i18n@node_name",
+					"value": "审批"
+				},
+				{
+					"key": "@i18n@widget1",
+					"value": "Issue"
+				},
+				{
+					"key": "@i18n@widget2",
+					"value": "Stage"
+				},
+				{
+					"key": "@i18n@widget3",
+					"value": "Description"
+				}
+			]
+		}
+	],
+	"node_list": [
+		{
+			"id": "START"
+		},
+		{
+			"id": "approve-here",
+			"name": "@i18n@node_name",
+			"approver": [
+				{
+					"type": "Free"
+				}
+			]
+		},
+		{
+			"id": "END"
+		}
+	],
+	"viewers": [
+		{
+			"viewer_type": "NONE"
+		}
+	]
+}`
+	cancelExternalApprovalReq = `{
+  "approval_code": "%s",
+  "instance_code": "%s",
+  "user_id": "%s"
+}`
+)
+
+// GetTenantAccessToken gets tenant access token.
+func (p *feishuProvider) GetTenantAccessToken(appID string, appSecret string) (*TenantAccessTokenResponse, error) {
```

Comment:
This method doesn’t need to be public. Even though we don't have an interface, let's still be crystal clear about what should be public and what not. 

More importantly, we need something similar to a token refresher. The token should be refreshed upon expiration error.

https://github.com/bytebase/bytebase/blob/main/plugin/vcs/gitlab/gitlab.go#L295